pkg/workloads/cortex/lib/client/tensorflow.py

```diff
@@ -12,320 +12,498 @@
-            # so it also gets to be tested against its value
-            shape = "unknown"
-        elif input_metadata["tensorShape"].get("dim", None):
-            # known rank and known/unknown shape
-            shape = [int(dim["size"]) for dim in input_metadata["tensorShape"]["dim"]]
-        else:
-            raise UserException(
-                "invalid 'tensorShape' specification for input '{}' in signature key '{}' for model '{}'",
-                input_name,
-                signature_key,
-                model_name,
-            )
+        return "loaded tensorflow model"
 
-        parsed_signature[input_name] = {
-            "shape": shape if type(shape) == list else [shape],
-            "type": DTYPE_TO_TF_TYPE[input_metadata["dtype"]].name,
-        }
-    return signature_key, parsed_signature
+    def _remove_models(self, model_ids: List[str]) -> None:
+        """
+        Remove models from TFS.
+        Must only be used when caching enabled.
+        """
+
+        models = {}
+        for model_id in model_ids:
+            model_name, model_version = model_id.rsplit("-", maxsplit=1)
+            if model_name not in models:
+                models[model_name] = [model_version]
+            else:
+                models[model_name].append(model_version)
 
+        model_names = []
+        model_versions = []
+        for model_name, versions in models.items():
+            model_names.append(model_name)
+            model_versions.append(versions)
 
-def create_prediction_request(signature_def, signature_key, model_name, model_input):
-    prediction_request = predict_pb2.PredictRequest()
-    prediction_request.model_spec.name = model_name
-    prediction_request.model_spec.signature_name = signature_key
+        self._client.remove_models(model_names, model_versions)
 
-    for column_name, value in model_input.items():
-        if signature_def[signature_key]["inputs"][column_name]["tensorShape"] == {}:
-            shape = "scalar"
-        elif signature_def[signature_key]["inputs"][column_name]["tensorShape"].get(
-            "unknownRank", False
-        ):
-            # unknownRank is set to True if the model input has no rank
-            # it may lead to an undefined behavior if unknownRank is only checked for its presence
-            # so it also gets to be tested against its value
-            shape = "unknown"
+    def _determine_model_signature_key(self, model_name: str) -> Optional[str]:
+        """
+        Determine what's the signature key for a given model from API spec.
+        """
+        if self._models_dir:
+            signature_key = self._api_spec["predictor"]["models"]["signature_key"]
         else:
-            shape = []
-            for dim in signature_def[signature_key]["inputs"][column_name]["tensorShape"]["dim"]:
-                shape.append(int(dim["size"]))
+            signature_key = self._spec_models[model_name]["signature_key"]
```

Comment:
This can always be used right? Even if `self._models_dir == True`?

Reply:
What do you mean by always? It can be used in the context of the MMC when `self_models_dir` is False. It could also be accessed when `self._models_dir` is set to True, but isn't advisable.

Reply:
Simplified the method.